fastchat/serve/monitor/elo_analysis.py

```diff
@@ -399,6 +399,117 @@ def outlier_detect(
     return battles
 
 
+def fit_mle_elo(X, Y, models, indices=None, SCALE=400, INIT_RATING=1000):
+    from sklearn.linear_model import LogisticRegression
+
+    p = len(models.index)
+
+    lr = LogisticRegression(fit_intercept=False)
+    if indices:
+        lr.fit(X[indices], Y[indices])
+    else:
+        lr.fit(X, Y)
+
+    elo_scores = SCALE * lr.coef_[0] + INIT_RATING
+    # calibrate llama-13b to 800 if applicable
+    if "mixtral-8x7b-instruct-v0.1" in models.index:
+        elo_scores += 1114 - elo_scores[models["mixtral-8x7b-instruct-v0.1"]]
+    return (
+        pd.Series(elo_scores[:p], index=models.index).sort_values(ascending=False),
+        lr.coef_[0][p:],
+    )
+
+
+def construct_style_matrices(
+    df,
+    BASE=10,
+    apply_ratio=[1, 1, 1, 1],
+    style_elements=[
+        "sum_assistant_a_tokens",
+        "header_count_a",
+        "list_count_a",
+        "bold_count_a",
+        "sum_assistant_b_tokens",
+        "header_count_b",
+        "list_count_b",
+        "bold_count_b",
+    ],
```

Comment:
should we keep this list of style elements somewhere else not as default argument?

Reply:
What do you mean by keeping it somewhere?

Reply:
ah like maybe we should not set it as default argument, but initiate it when we call the function 
```
style_elements = [
      "sum_assistant_a_tokens",
      "header_count_a",
      "list_count_a",
      "bold_count_a",
      "sum_assistant_b_tokens",
      "header_count_b",
      "list_count_b",
      "bold_count_b",
]
X, Y, models = construct_style_matrices(battles, style_elements=style_elements)
```